chore: add backed reader, writer and pipe implementation

ibetitsmike opened the pull request:
Fixes: https://github.com/coder/coder/issues/19977

This PR introduces a new `backedpipe` package that provides reliable bidirectional byte streams over unreliable network connections. The implementation includes:

- `BackedPipe`: Orchestrates a reader and writer to provide transparent reconnection and data replay
- `BackedReader`: Handles reading with automatic reconnection, blocking reads when disconnected
- `BackedWriter`: Maintains a ring buffer of recent writes for replay during reconnection
- `RingBuffer`: Efficient circular buffer implementation for storing data

The package enables resilient connections by tracking sequence numbers and replaying missed data after reconnection. It handles connection failures gracefully, automatically reconnecting and resuming data transfer from the appropriate point.

ibetitsmike commented:
* **#19147** <a href="https://app.graphite.dev/github/pr/coder/coder/19147?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/coder/coder/19147?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

ibetitsmike committed:
"chore: add backed reader, writer and pipe"

ibetitsmike committed:
improvement to backed pipe concurrency and tests

ibetitsmike committed:
chore: moved backedpipe to be a part of the agent code

spikecurtis commented on agent/immortalstreams/backedpipe/backed_writer.go:
```diff
@@ -0,0 +1,244 @@
+				bw.onError(err)
+			}
+		}
+	}
```
Writes need to block while we are disconnected. Otherwise, the reliable connection doesn't get any backpressure and can just sling tons of data at us that will blow out the ring buffer.

spikecurtis reviewed (commented):
Not done with my review, but taking a break and wanted to send you comments so far.

spikecurtis reviewed (commented):
Another batch of comments.

I haven't really looked at BackedPipe yet, but I think maybe it would be better to wait for the comments I've made so far to be resolved.

ibetitsmike committed:
PR feedback implemented

ibetitsmike committed:
fmt

ibetitsmike committed:
revereted WaitSuperShort - not used anymore

ibetitsmike committed:
changed err to EOF instead of ErrClosedPipe and changed backed_writer to block on errors

ibetitsmike commented on agent/immortalstreams/backedpipe/backed_writer_test.go:
```diff
@@ -0,0 +1,411 @@
+	writer := newMockWriter()
+	err := bw.Reconnect(10, writer) // Future sequence
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "future sequence")
```
Fixed. Left them as before in ringBuffer since we don't export it.

ibetitsmike committed:
added sentinel errors

ibetitsmike committed:
fixed eviction tests in writer and moved from assert to require in reader tests

ibetitsmike committed:
added holding the mutex during reconnection and test cases to verify being closed/reconnected race conditions

ibetitsmike committed:
writer's tets cleanup

ibetitsmike committed:
improvements to backed pipe

spikecurtis commented on agent/immortalstreams/backedpipe/ring_buffer_internal_test.go:
```diff
@@ -0,0 +1,264 @@
+	if runtime.GOOS == "windows" {
+		// Don't run goleak on windows tests, they're super flaky right now.
+		// See: https://github.com/coder/coder/issues/8954
+		os.Exit(m.Run())
```
If Windows CLI tests are still flaky, then we need to look into it. That issue is closed not planned and has been for years.

spikecurtis commented on agent/immortalstreams/backedpipe/backed_pipe.go:
```diff
@@ -0,0 +1,290 @@
+	closed := bp.closed
+	bp.mu.RUnlock()
+
+	if closed {
```
Same deal as read --- the `writer` will return an error when it's closed, so you don't need to check the pipe. Also, writes don't return `EOF`.

spikecurtis commented on agent/immortalstreams/backedpipe/backed_pipe.go:
```diff
@@ -0,0 +1,290 @@
+	mu          sync.RWMutex
+	reader      *BackedReader
+	writer      *BackedWriter
+	reconnectFn ReconnectFunc
```
In order to handle both initial connections and reconnections, the thing that handles the connections is going to need to be stateful. On the client side it needs to post a new immortal stream the first time, and then store the name/ID of the stream for re-dialing.  Instead of making this a function, make it an interface type with the reconnection function on it.

It's much easier to understand this kind of thing because you can ask your IDE to show you all implementations of an interface type.

It also avoids the awkward closure that you define up the stack to handle the statefulness. It can be a regular function on the struct type rather than an anonymous closure over it. Also makes stack traces way more readable.

ibetitsmike commented on agent/immortalstreams/backedpipe/backed_writer.go:
```diff
@@ -0,0 +1,233 @@
+		// Notify parent of error
+		select {
+		case bw.errorChan <- err:
+		default:
```
The reconnect process is debounced by singleflight on the pipe level so I wouldn't expect them to trigger two reconnects. 

However I've added separate channels for read & write errors in the backed pipe.

Is comment describing the behavior enough or do you feel there might be an edge case here I'm not testing for right now?

ibetitsmike replied:
We set the writer to nil so no new errors can appear since we block until we reconnect. With the improved handling of concurrent reader/writer errors this can now be closed.

ibetitsmike committed:
coordinating tests with channels
separate channels for writer/reader errors
moved to an interface for reconnection
coordination logic for closing a pipe

ibetitsmike committed:
added connection state to backed_pipe and connection generation for better error tracking

ibetitsmike committed:
fixed flaky tests

spikecurtis commented on agent/immortalstreams/backedpipe/backed_pipe.go:
```diff
@@ -0,0 +1,346 @@
+func NewBackedPipe(ctx context.Context, reconnector Reconnector) *BackedPipe {
+	pipeCtx, cancel := context.WithCancel(ctx)
+
+	errorChan := make(chan ErrorEvent, 10) // Buffered for async error reporting
```
What's the deal with 10? Like, why is that the right size for this buffer?

In general, I'd encourage you to consider anything other than 0 or 1 to be suspicious. There are legit uses for other numbers, but generally only if we are "buffering" to smooth out spiky usage.

spikecurtis commented on agent/immortalstreams/backedpipe/backed_reader_test.go:
```diff
@@ -0,0 +1,612 @@
+	}()
+
+	// Wait for read to start
+	testutil.RequireReceive(ctx, t, readStarted)
```
`readStarted` doesn't actually verify that the `Read` call has started, since it happens on line 121 and read doesn't happen until line 123. As such it doesn't really add anything to this test and should get removed.

spikecurtis commented on agent/immortalstreams/backedpipe/backed_reader_test.go:
```diff
@@ -0,0 +1,612 @@
+			// Read is still blocked, which is what we want
+			return !br.Connected()
+		}
+	}, testutil.WaitShort, testutil.IntervalMedium)
```
I don't understand why this is wrapped in `Eventually` --- the reader has never connected at this point, so `br.Connected() == true` should be a test failure, not a retry. The other branch of the select is also a test failure.

ibetitsmike committed:
PR review fixes

spikecurtis commented on agent/immortalstreams/backedpipe/backed_pipe.go:
```diff
@@ -0,0 +1,357 @@
+	// - readerSeqNum: how many bytes we've successfully read from remote (send to remote)
+	// - writerSeqNum: how many bytes we've written so far (used for validation)
+	readerSeqNum := bp.reader.SequenceNum()
+	writerSeqNum := bp.writer.SequenceNum()
```
If you read these values here, they can be out of date by the time you use them, since you release the locks and reads & writes can flow.

spikecurtis commented on agent/immortalstreams/backedpipe/backed_pipe.go:
```diff
@@ -0,0 +1,357 @@
+	}
+
+	// Validate sequence numbers
+	if remoteReaderSeqNum > writerSeqNum {
```
The writer's `Reconnect()` method already handles this check, and it does so while holding the lock. Doing it here means we are subject to data races which could give false errors.

ibetitsmike committed:
PR review fixes

ibetitsmike committed:
mock reconnector call count improvements
